### DISCUSSION TITLE
What alignment guarantees come from `CFG_TUSB_MEM_SECTION` and `CFG_TUSB_MEM_ALIGN`?

### DISCUSSION THREAD

**Discussion by bmbrownlee** [hathach/tinyusb#2260] [Q&A] [answered]
All the example configurations show following block on these two defines:

```c
/* USB DMA on some MCUs can only access a specific SRAM region with restriction on alignment.
  * Tinyusb use follows macros to declare transferring memory so that they can be put
  * into those specific section.
  * e.g
  * - CFG_TUSB_MEM SECTION : __attribute__ (( section(".usb_ram") ))
  * - CFG_TUSB_MEM_ALIGN   : __attribute__ ((aligned(4)))
  */
```

I assumed that these would guarantee that all accesses to variables in the section `CFG_TUSB_MEM_SECTION` would be aligned. This was perfect, since this meant I could put this section on a noncacheable memory page and forgo the overhead of needing to repeatedly flush and invalidate cache. However, a lot had to be done to get this to actually be the case. All over the code I had to add `volatile` specifiers to prevent `uint8_t` reads/writes from being optimized into unaligned `uint16_t` reads/writes, manually add padding to various structures, and write a new `memcpy`/`memset` wrapper to enforce alignment. It works now, but every once in a while a build spawns a new compiler optimization, and I have to go and fix the code.

So what alignment guarantees is `CFG_TUSB_MEM_SECTION` supposed to have? Seeing how many changes had to be done to get it to work, I have a feeling that I interpreted it wrong.

Some possibly relevant info:

- version: 0.15.0 (latest release)
- build platform: linux, using gcc
- target platform: custom / no os
- target board: i.MX 6 SoloLite (small tweaks from i.MX RT family)

**Comment by Rocky04** [comment]
**I would assume that the compiler should align all the variables normally to the standard of the used architecture and that it only wouldn't do that if the packed attribute is used. So normally it should already use padding to ensure the alignment of the specific data types depending of the architecture. On the other side a specific **minimum** alignment can be forced by using the attribute for it.**

For example this snipped should typically have a size of 12 bytes. So for optimization the elements should be better reordered by placing putting "c" between "a" and "b". In the case this order is needed without padding, like for transmitting it, the packed attribute should be used. Then the size would be 8 bytes.
```
typedef struct
{
        uint8_t   a;
        // One byte padding
        uint16_t  b;
        uint8_t   c;
        // Three bytes padding
        uint32_t  d;
} foo_t;
```

On a 32 bit architecture it can be that `uint_8` can be maybe put on any address, `uint_16` only on an even address while `uint_32` needs to be on an address divisible by four. A 16 bit MCU wouldn't be able to address a native 32 bit data element, these then typically utilize two registers if the MCU can still do 32 bit operations. Because then two register need to be loaded these would be still aligned individually as shorts and so both on an even address. So the whole 32 bit number would be aligned just by an even address _(in that case there should be no need that the data needs to be on an address divisible by four)_. In that case the padding between "c" and "d" should only be one byte instead of three.

I would assume that the section attribute wouldn't have any impact on the alignment. _As far as I understood it it should just put the data into a dedicated memory area / address section. Meaning the data would not be next to the other data of the same source file._

~~But I have no clue why the alignment is forced at some points in TinyUSB. I would think this would only make sense if that is used for a field in a packed structure (which would be weird) or if the data is used for a communication with a different architecture as is. _Like when structs are transmitted as RAW bytes._~~ _TinyUSB likely needs an alignment of four for some buffers when these are declared as byte arrays. Like a 32 bit MCU may copies whole integers and not single bytes because it's much faster to use the native size. In this case the buffer must be aligned that way. Otherwise the buffer address might not start on an address divisible by four and so data is may cut at the start or it starts with wrong data at the beginning while the same amount is cut at the end._ I had a similar issue when I used wide character strings as an easy initialization for complete string descriptors [here](https://github.com/Rocky04/tinyusb/blob/575acc5eeeb22931744a3a70dc0241754811d275/examples/device/hid_custom_class/src/usb_descriptors.c#L294). Because I don't wanted to generate the string descriptors on each request and instead just return pre-defined ones in order to save compute time in exchange of some additional flash memory usage.

_Edit:_
But yeah, the compiler would be allowed to extend the size of a variable to match it's native register size / architecture alignment. So for example to add three padding bytes for a char on a 32 bit architecture for a faster access. I don't really know how the instructions work on the hardware. For example as I understood it a x86 CPU has different instructions for the registers depending on the size which should be used. So for example RAX for 64 bits, EAX is for 32 bits, AX for 16 bits and AH & AL for 8 bits. So except AH which has an offset by eight bits all other should have no offset. So all these instructions address the same register "A" but with a different size while ignoring the left over upper bits. Meaning it can be that on a 32 bit architecture it may would not be possible to load the upper short of an integer directly. Imagine two shorts next to each other. So the first can be accessed and processed directly via the AX register while the second would be needed to be shifted to the right by 16 bits before that, in the case the data can only be loaded via a four byte alignment architecture.

Same for any boolean regardless of the actual architecture. The compiler can pack multiple together into an integer so save memory. But then to address these the compiler would need to add mask operations in order to read or write to a specific bit which would makes it much slower. Like to trade application speed in order for memory economy. So on a modern system where RAM isn't a problem anymore boolean are basically always extended to smallest possible size which can be accessed. _So typically at least a char._

**_tl;dr_**
- You would need to check the specific architecture and how the compiler handles these to know how the data is aligned and so how different data sizes can be addressed. _According to the product details the i.MX 6SoloLite supports 16/32 bit RAM. So I assume the smallest size which can be accessed is a short and so the compiler likely extend any data field to that if optimizations are enabled._
- I don't really get what you mean with unaligned `uint_16`. That would be a short starting at an uneven address which should only be possible if you used the packed attribute. I guess you mean that a char is padded and so takes the same space as a short. But then I'm confused why volatile would prevent this. I would imagined that it would be the other way around, that the compiler would not add padding when the value is cached / only hold in a register, but padded when it's always needed that this is loaded from memory.

**Reply by bmbrownlee** [reply]
Part of the problem is not necessarily what is legal or efficient on my architecture or chip, but what is legal in the memory section specified by `CFG_TUSB_MEM_SECTION`. In my case, unaligned accesses are completely legal (the compiler will use them as it deems efficient) as long as the MMU has been configured to allow them. However, for `CFG_TUSB_MEM_SECTION` I am using a noncacheable memory page, and all noncacheable memory pages can only have aligned accesses.

There is a bit of a difference between the alignment of structures and their member variables versus the reads/writes that the compiler ends up doing. For example, in `cdcd_init()`  in `cdc_device.c`, there is a part there is this block of code:

```c
cdcd_interface_t* p_cdc = &_cdcd_itf[i];

p_cdc->wanted_char = (char) -1;

// default line coding is : stop bit = 1, parity = none, data bits = 8
p_cdc->line_coding.bit_rate  = 115200;
p_cdc->line_coding.stop_bits = 0;
p_cdc->line_coding.parity    = 0;
p_cdc->line_coding.data_bits = 8;
```

Since `_cdcd_itf` is put in `CFG_TUSB_MEM_SECTION`, all reads and writes with `p_cdc` need to be aligned. The first thing that needs to be fixed is that two bytes of padding need to be added between `wanted_char` and `line_coding.bit_rate` so that `line_code.bit_rate` is four byte aligned for a four byte read. The rest of the writes are to bytes (which are always aligned), but there is a subtle problem still: we have not told the compiler they must be written as bytes. GCC will see that we are writing zero to both `stop_bits` and `parity` and sometimes decide that grouping those together as a single two byte unaligned write to be more efficient. Because of the noncacheable section, this leads to an immediate exception. The solution to this is to use `volatile` to force the compiler to access these members the way we want them to be.

These are the types of issues that I had to fix around the source code to get tinyusb to work for my needs.

**Comment by Rocky04** [comment]
As mentioned the compiler should do the needed padding normally. So because the first element of `line_coding` is declared as an integer the compiler should add the needed padding before if needed. The alignment attribute should only ensure the minimum boundary of it. So this attribute should not only specify the minimum size of an element but also the address of that depending on the size. So for example that `TU_ATTR_ALIGNED(4) uint_8 foo` would not only cause that foo takes up 4 bytes but also that it would cause that the address for that is divisible by four. **So you should check the compile options why this isn't the case.** You should also check how your compiler handles the `TU_ATTR_ALIGNED` macro and that this is supported as intended for your compiler and that the macro is correctly expanded by TinyUSB. For GCC it should be defined [here](https://github.com/hathach/tinyusb/blob/56c6d2feab2e6e68a539534910900ae9f4e593d1/src/common/tusb_compiler.h#L127). **Be aware that the alignment attribute isn't a compiler standard but rather a compiler extension. Meaning you need to enable these (by at least using c99 for gcc).**

The compiler only puts the data in a special section if you define the `CFG_TUSB_MEM_SECTION` attribute in your compile settings. On default no dedicated section is used ([see here](https://github.com/hathach/tinyusb/blob/56c6d2feab2e6e68a539534910900ae9f4e593d1/examples/device/cdc_msc/src/tusb_config.h#L78)). It's up to the linker to then put that at a special address range by providing him the corresponding mapping. So I would assume the compiler may don't know anything about the address range at compile time and so it would apply his default alignment _because only the linker would know but that comes later in the build process. So I assume that the macro `CFG_TUSB_MEM_SECTION` don't gives any alignment guarantee._

Because it seems you need two bytes of padding between the element `wanted_char` and `line_coding` and not three this tells me that `_cdcd_itf` starts with an address offset of one. So yeah, if you need to add the padding manually it would be a nightmare to do that when something has changed. Like every time you add or remove a byte in your code it can be that the address changes and so the padding need to be adjusted again. _By the way, the reason why the `cdc_line_coding_t` struct is used with the alignment attribute is likely because that has the packed attribute. The packed attribute can break the alignment but because it has an integer element which needs to be aligned to be accessed as a whole integer the structure need to be at an address divisible by four on a 32 bit architecture._

That the compiler initializes the `stop_bits` and `parity` element altogether seems to be a reasonable optimization for me. If both are set at the same time this will save some overhead in addressing them. I don't understand why this should cause any issues. _Because the struct where these elements are has the packed attribute the compiler wouldn't be allowed to add padding for them. Meaning he can't extend the size of the elements to a short for an independent and fast access._

I also have the feeling that my understanding of alignment differs from yours. Why do you think the elements `stop_bits` and `parity` are unaligned when the compiler writes to both at the same time? Again, it would be unaligned if the short data would be on an uneven address. But because the compiler writes to both at the same time it would indicate that this is aligned. Due to your manual padding you also should already ensure that these are indeed aligned. _Accessing two chars altogether via type punning of an operation of the short size isn't unaligned access. Alignment has only something to do with the address of data elements depending on their size._

_Edit:_
I also struggle what you mean with "noncacheable memory page". It don't make any sense to me why putting the volatile attribute to any element in that section would change anything. If you use volatile you tell the compiler to force to load the data from the memory address and so to don't use any cached value. Meaning the data should not be used from a register nor from the processor cache. I would assume if the memory page is non-cacheable it would be already the same as volatile and so that the volatile attribute would be obsolete. I'm also confused why you would need to repeatedly flush and invalidate the cache if you aren't using a non-cacheable memory page to begin with. The i.MX 6SoloLite should be a single core without SMT, so I don't get why you don't want to utilize the cache.

_Edit 2:_
Because the alignment don't seems to be there / working for you I wouldn't be suppressed if the issue you encounter is because "unaligned accesses are completely legal" for you. Like that you are maybe using an option which tells the compiler to ignore any alignment boundary and so causing all the issues for you.

**Reply by bmbrownlee** [reply]
The structure needs to be padded correctly. I actually just discovered that this was fixed somewhat recently (after most recent release) in [commit 190acc1](https://github.com/hathach/tinyusb/commit/190acc1fb0579d3e097ba2b0b96033740ecee26e). 

Unless I am mistaken, `volatile` has nothing to do with cache; it is just a type specifier that suggests to the compiler that the value could change at any point, forcing it to read/write it. This could be reading/writing from cache or DDR. It also has the benefit of specifying to read/write exactly this value, which prevents the compiler from optimizing it by bundling it with other reads/writes.

Yes, part of the issue is the fact that unaligned accesses are generally legal. Otherwise, the compiler would never generate code to do them. The problem is that unaligned accesses on my platform has pretty minimal (to almost none?) additional cost, so compiling with them allows the compiler to do a significant amount of optimization. I could try some parts of the code with/without alignment restrictions, but things become a bit fragile since code without alignment restrictions cannot call code with them.

**Reply by Rocky04** [reply]
- So basically I missed that you were you using the release version of the code where the alignment attribute was missing and you missed that I was referring that this should be present.

>By the way, the reason why the `cdc_line_coding_t` struct is used with the alignment attribute is likely because that has the packed attribute. The packed attribute can break the alignment but because it has an integer element which needs to be aligned to be accessed as a whole integer the structure need to be at an address divisible by four on a 32 bit architecture.

- No, the volatile attribute tells the compiler to always load the data from the memory address every time and so ignore **any** cache. _So that is basically all about non-caching._ Imagine an address space used for DMA or any I/O operation / register which is memory map. If the compiler would not ensure that the data is loaded again every time and instead cached data from a previous read would be returned it would cause issues. _Changing GPIO ports would may not work and also you would not be able to send data because only the last write would be performed._

- It's undefined behavior, some compiler throw an error others don't, on some systems it will just work on others it causes issues. Being aware of this and writing code which is truly universal isn't that easy. Either the data source should ensure alignment to begin with or the data consumer should assume a potential unaligned source and so need to copy the data into a temporary variable. _As far as I know the alignment attribute isn't part of the old C specification because that was created for a system which can handle misalignment just fine and there is a reason why this is an extension._

**Comment by hathach** [comment, marked as answer]
> So what alignment guarantees is CFG_TUSB_MEM_SECTION supposed to have? Seeing how many changes had to be done to get it to work, I have a feeling that I interpreted it wrong.

CFG_TUSB_MEM_SECTION/CFG_TUSB_MEM_ALIGN allow tinyusb to place endpoint memory in specific section/address with specific alignment, which is required for some mcu such as NXP imxrt/lpc55 which require all buffered that is used with USB DMA must reside on a specific address (on some mcu such as lpc1778) with alignement of 64. 

Therefore CFG_TUSB_MEM_SECTION has no impact on the alignment, just for placing it in specific section if needed. But the **CFG_TUSB_MEM_ALIGN** does, and it is used only for memory/buffer that is used to for making USB transfer.

**Reply by bmbrownlee** [reply]
Thanks, this is exactly what I wanted to know. I guess the problem is that there is a mismatch between CFG_TUSB_MEM_SECTION and CFG_TUSB_MEM_ALIGN; not everything placed in the memory section is guaranteed to accessed with a particular alignment. This would mean that either I would need to stop using the noncacheable memory section (which has the alignment restrictions), or maintain my own version. Or, if anyone else wants to run tinyusb in this manner, I could draft a PR for it.

**Reply by hathach** [reply]
latest master includes the dcache support for imxrt (flush/invalidate), you would also need to define CFG_TUSB_MEM_SECTION=32 (min cache line). If it still does not work with your port, try to define non-cachable section and have CFG_TUSB_MEM_SECTION with attribut((section())) to place them there accordingly.

**Comment by Rocky04** [comment]
> I guess the problem is that there is a mismatch between CFG_TUSB_MEM_SECTION and CFG_TUSB_MEM_ALIGN; not everything placed in the memory section is guaranteed to accessed with a particular alignment.

There is no mismatch between them. As mentioned multiple times now the section attribute specifies in which section the compiler should place the data (for the linker) and so how the data is grouped together and the alignment specifies the boundary (the starting address) of the data. These are two different things...

As uninitialized data `_cdcd_itf` should be normally placed in the .bss section together with all the other uninitialized data of the other source files. _If it would be initialized it would be part of the .data section instead._ With `CFG_TUSB_MEM_SECTION` you can use a custom section. _But in order to actually use a specific memory range you would need to tell the linker where to map these._ But even organizing the data as you want can help to maybe debug it better.

About the alignment, a modern x86 CPU should not care about misalignment but an ARM MCU does as far as I know. There is a reason why the attribute is a gcc extension and not part of an older C standard. As far as I know the `_Alignas` keyword was added with C11 and `alignas` with C23.

_Edit:_
The issue for you seem to be the missed forced alignment attribute. That was needed to restore the correct alignment because that was removed by the packed attribute of the structure.

**Comment by Rocky04** [comment]
@hathach 
> ...must reside on a specific address (on some mcu such as lpc1778) with alignement of 64.

I'm curious about the alignment of LCP1778. Do you really meant that this need an alignment of 64 or was it a typo? As a 32 bit MCU I would expect the alignment to be 4 and that this is even the upper limit. It don't make any sense to me why that MCU should align anything on an address divisible by at least 64. So this would mean a boundary of 64 byte. Meaning it would align to the current biggest SIMD instructions available. _Like a complete register (ZMMx, YMMx & XMMx) for AVX-512 is 64 bytes wide._ Sadly I haven't found anything related to the alignment of that MCU in the docs.

**Reply by HiFiPhile** [reply]
Some USB IPs need special alignment handing, many LPC need 64 bytes alignment.

In this case the phy assume buffer's address[5..0] are 0.
![image](https://github.com/hathach/tinyusb/assets/4375114/28daaa8a-4429-4d00-bb91-d318040358e7)

**Reply by Rocky04** [reply]
@HiFiPhile 
Thx.. It would be nice if you can link the source so I can read it later on completely. So far I can't read anything in the extract which refers to an alignment of 64 bytes. I don't see why the address offset and the package count (with a size of 64 byte) should have anything to do with the alignment of the used buffer. I doubt that the hardware actually would write 64 bytes at once and so need that alignment. How can it be if a register size is only 32 bits wide and if the data from USB only comes sequentially?

_Edit:_
Using blocks of 64 byte to copy data isn't the same as to align each of these blocks to a boundary of their size. ~~From my research, the alignment has an upper limit and is capped by that. Meaning I assume the actual alignment is reduced to 4.~~ I would be very surprised if that isn't the case. But there is no way for me to prove that because I don't have such a board to inspect the address of a buffer which has this high alignment attribute.

_I will check the cap of the alignment value on a GD32 board. If I'm wrong it would let me use greater values than the register is wide. While I doubt that the mentioned MCU would allow an alignment of 64 bytes, that is hardware dependent. So if the alignment is capped for me it would not be a 100% proof that this is also the case for the other MCU. Even it's unrealistic..._ :-)

**Correction...**
So I checked the alignment and it don't seem to be capped. So I either picked it up wrongly or my source wasn't good.

**Reply by HiFiPhile** [reply]
> It would be nice if you can link the source so I can read it later on completely. 

You can get UM11126 from NXP, chapter 44.7

> I doubt that the hardware actually would write 64 bytes at once and so need that alignment. How can it be if a register size is only 32 bits wide and if the data from USB only comes sequentially?

Without having the HDL source it's difficult to know the reason. But usually in hardware world this kind of thing can simplify HDL expression thus save some logic gates. ( ~~Like some old 64bit XEON only has 36bit address line~~ )

**Reply by Rocky04** [reply]
- It is reasonable that if the size always changes in a constant way, like by 64 bytes, that the first six bits can be omitted because it wouldn't make any sense to provide them. It would be only a waste of space to provide hard coded data which would never ever change. But this don't mean that the buffer of the application need to live on these address boundaries and so that these address bits for the buffer must be also zero.
- The USB data handling only seems to work in 64 byte blocks there. So RAM is wasted in the case a package is smaller than one block in favor for some optimizations and simplifications. So the only requirement is that the buffer is a multiple of 64 bytes big to ensure that  complete package can be hold. But it don't mean necessarily that it needs to be aligned that way. **_Sadly it is because the base address of the buffer which is used together with the offset also don't holds the lower bits._**
- The text even mentions that this is an offset, meaning it can be an offset from any arbitrary address. Imagine a buffer of 512 bytes. That would be able to hold up to 8 packages of 64 bytes. The buffer and so the first package may start at the address 0x20000114. Then the second package would start with an offset of 64 bytes (0x40) compared to that and so at address 0x20000154, the third at 0x20000194 and so on. So depending how the (DMA) transfer is configured it should be fine to even be byte aligned while 4 byte aligned would be more efficient. So the first six bits should be kept untouched and the buffer moves by 64 bytes each transfer. That the DMA transfer supports burst mode should also not change anything. _**Sadly there is no way to store a full buffer address with all bits.**_

- The documentation of the above mentioned LCP1778 states "Supports 8, 16, and 32-bit wide transactions" for the general purpose DMA controller. Meaning it should support byte alignment as well as 4 byte alignment but not any more. _So even if there is a special internal buffer with an alignment higher than 4 bytes the DMA transfer would reduce that to 4 if that handles the copying._ _**Problem is the reduction in the address resolution, so it's a lack of precision and not alignment.**_
https://www.nxp.com/docs/en/data-sheet/LPC178X_7X.pdf

- About the UM11126, the manual only mentions 8 bit, 16 & 32 bit transfers respectively 1 byte, 2 bytes and 4 bytes alignment in the section "22.5.7 Address alignment for data transfer" for the DMA controller (section 22.5).
- Both the USB0 FS host (section 42.7) and USB1 HS host (section 43.5) mention that "All registers are 32 bits wide and aligned in the word address boundaries." _So if the data would be handled by a USB register the alignment would be limited to 4 at max too._ 

~~=> So only if the data is stored in a special memory area which is aligned bigger than a 4 byte boundary and if that data is not copied via DMA but rather done via some wider special instructions, it would be possible to have a wider alignment as 4 bytes.~~

**Reply by Rocky04** [reply]
**I now read through some parts of the document and can't really believe what I read there.** I'm shocked in the way how it seems to be implemented (and documented)...

So yeah, it looks like the transfer buffer needs indeed to live at an address which is aligned to a 64 byte boundary so that  the USB module reads and writes at the correct position. _So it's due to a lack of address resolution / precision and not the alignment per se._ But I'm still not really sure if the complier is able to ensure this boundary. So either my assumption what I found about a capped upper alignment isn't true **(and yes it seems to be false)** when I learned about this or somehow the compiler would need to know that this big alignment is possible for these MCUs **(and so the compiler seems to not care)**.

But I'm speechless about:
- There is not a single clear mentioning that the buffer for the endpoint needs to be aligned in a 64 byte boundary because the USB module assumes a specific address number to read from and write to. I think this alone is ridicules.
- The composition of the transfer buffer consist of a base address of the "Data Buffer Start Address" (bits 32:22), the individual endpoint address offset (21:6) and a fixed lower address of zero bits (5:0). [41.4.1 USB0 software interface](https://www.mouser.com/pdfDocs/NXP_LPC55S6x_UM.pdf#page=787&zoom=100,66,105), [44.4.1 USB1 software interface](https://www.mouser.com/pdfDocs/NXP_LPC55S6x_UM.pdf#page=854&zoom=100,66,105) So 22 bits are unused where the base address is specified. I was expecting that there the entire starting address would be defined which is then just shifted by the 64 byte blocks depending on their amount. To me this seems to be an arbitrary limitation. But even if there is an argument that this would need a calculation of the address due to shifting by the amount of packages and so would not be efficient anymore there is still a better way. It would be reasonable to also define the first six bits of the starting address there and so only have the middle section as zero bits, so that the stating address and offset still can be combined by a simple or-operation and so maintaining a complete 32 bit address without any constant bits parts. 
- USB0 FS device allows to set the upper bits of the buffer start address. [41.7.3 USB0 EP command/status list start address](https://www.mouser.com/pdfDocs/NXP_LPC55S6x_UM.pdf#page=793&zoom=100,66,105) While for USB1 HS device that is fixed and read only. [44.6.4 USB1 data buffer start address](https://www.mouser.com/pdfDocs/NXP_LPC55S6x_UM.pdf#page=861&zoom=100,66,105) It feels abstruse to allow it for the slower speed but not the faster one. _Possible that the document is wrong and the address is also fixed for FS, who knows..._
- There is a dedicated USB (S)RAM section which is limited to 16 KB. [1.1 Introduction](https://www.mouser.com/pdfDocs/NXP_LPC55S6x_UM.pdf#page=4&zoom=100,66,105) If this section must be used for USB it makes no sense to me to even have a starting buffer address at all. The offset for this fixed section can be accessed via 14 bits and with a predefined 4 byte alignment only 12 bits would be needed which is less than the currently 16 bits which are used for the individual endpoint offset. And in the case the buffer can live anywhere in the memory I don't get why even having a dedicated USB (S)RAM section to begin with. _A 32 bit value for the buffer address only makes sense if the MCU can utilize external RAM._

_I never ever imagined that a company which spends tons of money to develop such a chip would do it that way._
